### PR TITLE
Changes to ensure concurrent requests modify the same session

### PR DESCRIPTION
**Pull request by alxdarksage**
JIRA https://sagebionetworks.jira.com/browse/BRIDGE-2342

I am able to reproduce BRIDGE-2342 now with a JMeter script that runs the two requests concurrently. Interestingly, the changes mean that the second sign in, which is retrieving activities, will now return 401, "Not signed in," once the first consent succeeds, because it will overwrite the session token that was issued to request two. Whereas before it would keep returning 412. That's despite the fact that the login for request #2 just succeeded.

This behavior is an improvement because it will retriever a re-sign in rather than a consent flow of some kind.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/cache/CacheProvider.java [hunk 1]
```diff
@@ -127,24 +114,92 @@ public void setUserSession(final UserSession session) {
...
+        UserSession session = getUserSessionNewVersion(sessionToken);
+        if (session == null) {
+            session = getUserSessionOldVersion(sessionToken);
+        }
+        return session;
+    }
+    
+    private UserSession getUserSessionNewVersion(String sessionToken) {
+        checkNotNull(sessionToken);
+        try {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/cache/CacheProvider.java:129]: Do we need unit tests for the new get methods? Should include the following cases:

* session token not found
* user ID has no session
* session ID mismatch
* session successfully retrieved

File: app/org/sagebionetworks/bridge/cache/CacheProvider.java [hunk 2]
```diff
@@ -127,24 +114,92 @@ public void setUserSession(final UserSession session) {
...
+    public UserSession getUserSessionByUserId(String userId) {
+        UserSession session = getUserSessionByUserIdNewVersion(userId);
+        if (session == null) {
+            session = getUserSessionByUserIdOldVersion(userId);
+        }
+        return session;
+    }
+    
+    private UserSession getUserSessionByUserIdNewVersion(final String userId) {
+        checkNotNull(userId);
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/cache/CacheProvider.java:185]: Similarly, we should add unit tests for this as well, including the following cases:

* user ID has no session
* user ID has session

File: test/org/sagebionetworks/bridge/cache/CacheProviderMockTest.java [hunk 3]
```diff
@@ -171,39 +162,52 @@ public void testGetUserSessionByUserId() throws Exception {
...
+        session.setSessionToken(DECRYPTED_SESSION_TOKEN);
         
         cacheProvider.removeSession(session);
-        cacheProvider.getUserSession(SESSION_TOKEN);
+        cacheProvider.getUserSession(DECRYPTED_SESSION_TOKEN);
         
-        verify(transaction).del(SESSION_KEY.toString());
-        verify(transaction).del(USER_SESSION_KEY.toString());
+        verify(transaction).del(TOKEN_TO_USER_ID.toString());
+        verify(transaction).del(USER_ID_TO_SESSION.toString());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/cache/CacheProviderMockTest.java:222]: Test should verify that we delete all 4 keys.

File: test/org/sagebionetworks/bridge/cache/CacheProviderMockTest.java [hunk 4]
```diff
@@ -171,39 +162,52 @@ public void testGetUserSessionByUserId() throws Exception {
...
+        String ser = BridgeObjectMapper.get().writeValueAsString(session);
+
+        when(jedisOps.get(USER_ID_TO_SESSION.toString())).thenReturn(ser);
+        
         cacheProvider.removeSessionByUserId(USER_ID);
         
-        verify(transaction).del(SESSION_KEY.toString());
-        verify(transaction).del(USER_SESSION_KEY.toString());
+        verify(transaction).del(TOKEN_TO_USER_ID.toString());
+        verify(transaction).del(USER_ID_TO_SESSION.toString());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/cache/CacheProviderMockTest.java:238]: Similarly, this should verify we delete all 4 keys.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/cache/CacheProvider.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/cache/CacheProvider.java:129]: I added tests for these specifically against the new getSession method alone. Will adjust once we cleanup the old methods.

File: test/org/sagebionetworks/bridge/cache/CacheProviderMockTest.java (hunk 3 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/cache/CacheProviderMockTest.java:222]: Added

File: test/org/sagebionetworks/bridge/cache/CacheProviderMockTest.java (hunk 4 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/cache/CacheProviderMockTest.java:238]: Added

**Review by DwayneJengSage — APPROVED** [review]

**Pull request merged**